### ISSUE TITLE
Write.configs error with ED2

### ISSUE THREAD

**Issue by serbinsh** [PecanProject/pecan#2252]
### Bug Description
I have run into a new bug with ED2 when trying to run with Fluxnet2015 drivers.  The met-process creates the drivers fine, though its a bit trial-and-error when attempting to figure out which met years it can successfully gap-fill.  However after the meta-analysis step and at the start of the write.configs step, it looks like PEcAn is not properly inserting the start/end MET dates into the met tags in pecan.METProcess.xml, which is leading to:

```
                                                                                                                                    113,4         53%
  NAs introduced by coercion
Warning in lapply(values_list[numeric_values], as.numeric) :
  NAs introduced by coercion
Warning in lapply(values_list[numeric_values], as.numeric) :
  NAs introduced by coercion
Warning in lapply(values_list[numeric_values], as.numeric) :
  NAs introduced by coercion
Error in proc_met_startend(settings[[c("run", "site", "met.start")]],  :
  length(rawval) == 1 is not TRUE
Calls: <Anonymous> ... do.call -> write.config.ED2 -> proc_met_startend -> stopifnot
```

Here is the met process xml

```
 </model>
 <run>
  <site>
   <id>676</id>
   <comment></comment>
   <name>Willow Creek (US-WCr)</name>
   <lat>45.805925</lat>
   <lon>-90.07961</lon>
  </site>
  <inputs>
   <css>
    <id>2000000005</id>
    <path>/data/Model_Data/sites/US-WCr//US-WCr.Inv2011.lat45.5lon-90.5.css</path>
   </css>
   <pss>
    <id>2000000006</id>
    <path>/data/Model_Data/sites/US-WCr//US-WCr.Inv2011.lat45.5lon-90.5.pss</path>
   </pss>
   <site>
    <id>2000000007</id>
    <path>/data/Model_Data/sites/US-WCr//US-WCr.Inv2011.lat45.5lon-90.5.site</path>
   </site>
   <met>
    <source>Fluxnet2015</source>
    <output>SIPNET</output>
    <username>sserbin</username>
    <id>
     <id1>2000000282</id1>
    </id>
    <path>
     <path1>/data/home/sserbin/.pecan/dbfiles/Fluxnet2015_ED2_site_0-676/ED_MET_DRIVER_HEADER</path1>
    </path>
   </met>
```

It is properly linking to the met file header, however, which is good.

In general, based on my experience lately, we need to do much testing and code-hardening on met-process/write.configs portion by model

Has anyone else run into this issue?  Hard to know if this is specific to Fluxnet given that I am having all sorts of issues with met process and AMF-LBL

**Comment by serbinsh** [comment]
Doh, I just looked back at a recent successful run with Sipnet using Fluxnet and the met process XML also doesnt have the met.start/end dates but it ran fine, so might need to see whats up with the ed wrapper.  Nor does the pecan.CHECKED version.....

**Comment by serbinsh** [comment]
@ashiklom I only tagged your here since you have been using ED2 a lot lately and wondered if you might have run into this before?  I suspect there is a check for the start/end that we 1) either dont need or 2) we could satisfy by writing the start/end dates into the XML via PEcAn

**Comment by serbinsh** [comment]
As suspected, re-running and putting in:

```

  <run>
    <site>
      <id>676</id>
      <met.start>1999-01-01</met.start>
      <met.end>2008-12-31</met.end>
    </site>
    <inputs>
      <css>
        <id>2000000005</id>
      </css>

```

avoided the original error.  But if you are generating met from scratch you dont want to have to know ahead of time what the final start/end met dates are.  It should be smart enough to insert these for you at the appropriate time.

**Comment by ashiklom** [comment]
AFAICT, `met.start` and `met.end` are only set [in the web interface](https://github.com/ashiklom/pecan/blob/eec8a63ea91b734efd5faf82c256f3ed62663a41/web/04-runpecan.php#L388-L390) -- we don't have any R code that sets them automatically.

Cue my rant about all the hidden stuff that the web interface does...

**Comment by ashiklom** [comment]
> But if you are generating met from scratch you dont want to have to know ahead of time what the final start/end met dates are. It should be smart enough to insert these for you at the appropriate time.

I agree 100%. The default assumption of the workflow should be that `met.start`/`met.end` are the same as those of the run, unless specifically specified by the user.

As to _where_ that should be set, that depends on how unique ED is in having a met cycling utility (or something that actually uses the `met.start`/`met.end` tags).

_Personally,_ I think it may be a good idea to not use ED's met cycling at all, and if we're going to run models with cycled met, to generate that met explicitly on the PEcAn side, so it can be used with other models as well.

**Comment by mdietze** [comment]
> Cue my rant about all the hidden stuff that the web interface does...

@ashiklom the job of the web interface is to help set up the pecan.xml for you. In that sense, EVERYTHING is does is 'hidden' by design.  If you don't use the web interface to set up your pecan.xml, why is it the web interfaces fault if you forget to include tags?

> if you are generating met from scratch you dont want to have to know ahead of time what the final start/end met dates are

I don't follow. If you're processing met, you sure has hell have to tell the system what dates you want processed. Why wouldn't you have to know that ahead of time?

**Comment by serbinsh** [comment]
AAAARRRRGGGHHHHH!!! The web strikes again!  That is it!  

As such, what we need to do now is 1) put in code to insert met.start/met.end for models like ED2 which will need it; 2) Perhaps create, as you say, the option in met process to recycle met for all models from the CF_gapfill(?) source.  For example, if you have 2000-2008 and you want to run 2000-2012, for the model-specific met record we could create say, Sipnet drivers that go 2000-2008 then 2009-2012 or similar.

@mdietze the issue is when you start a model run requesting new met then you are essentially setting the met start/end dates by setting the model run start/end.  As such when it gets to write.configs, should it not then insert these for you into the XML tags instead of crashing at WC?  That is, should the user instead have to put in met and model dates at the start before creating the met drivers?  I guess that is fine but it seems unnecessary 

To that end, if you point to a met ID in the database, should pecan not be able to look up that ID and insert the met dates for you as well?

**Comment by serbinsh** [comment]
To be fair, I can go either way its just sometimes hard to understand why sometimes we have things done for you in PEcAn and other times we need to be very specific about all of the settings/options, even when some of the info can come from the database.

**Comment by mdietze** [comment]
* The web interface assumes that run dates and met dates are the same and sets them as such, but allows you to change that default in Advanced Edit. I consider setting up a run that breaks that assumption to be a valid use of Advanced Edit.
* I don't think it makes sense for the met record in the DB to alter that default (met time = run time) because cycling met without the user's awareness is a very undesirable feature (e.g. I think I'm running 2018, but really I'm running 2001 met a second time, and then trying to validate against 2018 data?)
* I do think there's a sound argument that if met.start/met.end are not set manually or by the web, to have that set to the run dates during PEcAn.settings::prepare.settings
* If we do want to add met cycling within PEcAn, I do have some code around from PalEON that pseudo-'cycles' met by creating symlinks (e.g. 650.nc points to 1850.nc, 651.nc -> 1851.nc, etc). I didn't include that in the mainline because it seems dangerous to have the database think it has files back in the past that aren't really data, and it could also mess with updating (e.g. if set up a symlink with ERA5 met from 1978.nc -> 2017.nc, what happens when they back process 1950 to 1978 and PEcAn detects that it already has 1978?).

**Comment by ashiklom** [comment]
@mdietze My issue isn't with the automatic configuration of XML files, but rather that that configuration is distributed across two different systems -- PHP code in the web interface (which is in PHP and, more importantly, poorly documented) and R code in the `PEcAn.settings` package. In my opinion, and as I've said before, I think we should have every meaningful operation done in R, and have the web interface be a paper-thin wrapper that just calls some R code. In an ideal world, I would love to see a web interface completely written in R (via Shiny or something else). Alternatively (and much more low hanging fruit), I think many of the defaults that the web interface sets for you should just be set by the workflow script or settings package. To me, that would also facilitate the transition from beginner (web interface) user to intermediate user.

More generally, I agree 100% with @serbinsh 's point that it's hard to tell which user configured options are required and which are not. Partially, this is a documentation issue, but I tend to think that we should keep the minimum required user configuration as small as possible (especially if we can maintain provenance by spitting out the fully configured XML at the end, as we do).

**Comment by mdietze** [comment]
* There is a limit to what can be pushed to workflow.R and PEcAn.settings as we don't want default behaviors to override the ability to do Advanced Edit or manually configure the xml
* I'm fine with migrating the web GUI to shiny *in concept*, but in practice there are much more pressing issues that rewriting the GUI.
* A hybrid where we retain PHP for now but push things to R and have R write the pecan.xml seems a bit silly to me because the PHP will have to write the configs out to some other file format that R would then read in order to then write out the pecan.xml. To me this seems like it's just creating an unnecessary extra step.
* Yes, the documentation should be clear about what's [REQUIRED] and what's not. IMHO the pecan.xml documentation is absolutely required reading if you're doing anything beyond the basics, and should be a *very high priority* in fixing the documentation. Right now that doc doesn't indicate that met.start/end are required (it should until we change that), but those tags ARE included in the first 'minimal' example
* I do think this specific issue can be alleviated by adding the met.start/end = run start/end assumption into PEcAn.settings::prepare.settings for the case when met.start/end are not set explicitly. That would reduce the number of 'required' tags by 2.
